Services.Test/Concurrency/PerSecondCounterTest.cs

```diff
@@ -0,0 +1,246 @@
+
+            var target = new PerSecondCounter(MAX_SPEED, "test", this.targetLogger);
+
+            // Act
+            var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+            for (var i = 0; i < EVENTS; i++)
+            {
+                target.IncreaseAsync().Wait(TEST_TIMEOUT);
+            }
+
+            // Assert
+            var timepassed = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - now;
+            double actualSpeed = (double) (EVENTS - EVENTS_TO_IGNORE) * 1000 / timepassed;
+            log.WriteLine("Time passed: {0} msecs", timepassed);
+            log.WriteLine("Speed: {0} events/sec", actualSpeed);
+            Assert.InRange(actualSpeed, MAX_SPEED - 1, MAX_SPEED);
+        }
+
+        /**
+         * Test similar to "ItDoesntPauseWhenNotNeeded" and "ItObtainsTheDesiredFrequency_SeveralEventsPerSecond"
+         * The test runs 40 events in 20 seconds, the first 20 go through as a burst, then
+         * the caller is paused for 1 second, then the second half goes through as another
+         * burst. The test should take just a bit more than 1 second, definitely less
+         * than 2 seconds, so the actual speed should be ~39 events/second.
+         */
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void ItAllowsBurstOfEvents()
+        {
+            log.WriteLine("Starting test at " + DateTimeOffset.UtcNow.ToString("HH:mm:ss.fff"));
+
+            // Arrange
+            const int EVENTS = 40;
+            const int MAX_SPEED = EVENTS / 2;
+
+            var target = new PerSecondCounter(MAX_SPEED, "test", this.targetLogger);
+
+            // Act
+            var now = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
+            for (var i = 0; i < EVENTS; i++)
+            {
+                target.IncreaseAsync().Wait(TEST_TIMEOUT);
+            }
+
+            // Assert
+            var timepassed = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - now;
+            double actualSpeed = (double) EVENTS * 1000 / timepassed;
+            log.WriteLine("Time passed: {0} msecs", timepassed);
+            log.WriteLine("Speed: {0} events/sec", actualSpeed);
+            Assert.InRange(actualSpeed, EVENTS * 0.9, EVENTS);
+        }
+
+        /**
+         * Another "realistic" scenario, where 1 event happens every 250 msecs
+         * as if there was some I/O. Differently then other tests, this
+         * avoid bursts on purpose to make sure the internal logic of the
+         * rating logic is keeping the internal queue status correct.
+         */
+        [Fact, Trait(Constants.TYPE, Constants.UNIT_TEST)]
+        public void ItWorksWhenNoThrottlingIsNeeded()
+        {
+            // Arrange
+            var target = new PerSecondCounter(10, "test", this.targetLogger);
+
+            // Act
+            var paused = false;
+            for (var i = 0; i < 10; i++)
+            {
+                paused = paused || target.IncreaseAsync().Result;
```

Comment:
nit: could check for !paused and break earlier if you wanted to make the test return quicker.

Reply:
good point, I'll make the change